File: language/compiler/src/com/kubadziworski/parsing/visitor/expression/function/ParameterExpressionListVisitor.java
package com.kubadziworski.parsing.visitor.expression.function;

import com.google.common.collect.Lists;
import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.ParameterContext;
import jiva.parser.JivaParser.ParameterWithDefaultValueContext;
import jiva.parser.JivaParser.ParametersListContext;
import com.kubadziworski.domain.node.expression.Parameter;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by kuba on 09.05.16.
 */
public class ParameterExpressionListVisitor extends JivaBaseVisitor<List<Parameter>> {

    private final ExpressionVisitor expressionVisitor;

    public ParameterExpressionListVisitor(ExpressionVisitor expressionVisitor) {
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public List<Parameter> visitParametersList(@NotNull ParametersListContext ctx) {
        List<ParameterContext> paramsCtx = ctx.parameter();
        ParameterExpressionVisitor parameterExpressionVisitor = new ParameterExpressionVisitor(expressionVisitor);
        List<Parameter> parameters = new ArrayList<>();
        if(paramsCtx != null) {
            List<Parameter> params = Lists.transform(paramsCtx, p -> p.accept(parameterExpressionVisitor));
            parameters.addAll(params);
        }
        List<ParameterWithDefaultValueContext> paramsWithDefaultValueCtx = ctx.parameterWithDefaultValue();
        if(paramsWithDefaultValueCtx != null) {
            List<Parameter> params = Lists.transform(paramsWithDefaultValueCtx, p -> p.accept(parameterExpressionVisitor));
            parameters.addAll(params);
        }
        return parameters;
    }

}

File: language/compiler/src/parsing/visitor/expression/function/ParameterExpressionListVisitor.kt
package jiva.parsing.visitor.expression.function

import jiva.domain.node.expression.Parameter
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.ParametersListContext

/**
 * Created by kuba on 09.05.16.
 */
class ParameterExpressionListVisitor(private val expressionVisitor: ExpressionVisitor) :
    JivaBaseVisitor<List<Parameter>>() {

    override fun visitParametersList(ctx: ParametersListContext): List<Parameter> {
        val paramsCtx = ctx.parameter()
        val parameterExpressionVisitor = ParameterExpressionVisitor(expressionVisitor)
        val parameters: MutableList<Parameter> = ArrayList()
        if (paramsCtx != null) {
            val params = paramsCtx.map { it.accept(parameterExpressionVisitor)!! }
            parameters.addAll(params)
        }
        val paramsWithDefaultValueCtx = ctx.parameterWithDefaultValue()
        if (paramsWithDefaultValueCtx != null) {
            val params = paramsWithDefaultValueCtx.map { it.accept(parameterExpressionVisitor)!! }
            parameters.addAll(params)
        }
        return parameters
    }

}
